File: frontend/src/integrations/zoom/types/zoom.ts
export type ZoomAuthShape = 'oauth_user' | 'server_to_server'

export interface ZoomAccountSetupRequest {
  account_id: string
  display_name: string
  external_account_id: string
  account_email?: string | null
  metadata?: Record<string, unknown>
}

export interface ZoomLiveAccountSetupRequest extends ZoomAccountSetupRequest {
  auth_shape?: ZoomAuthShape
  client_id: string
  token_secret_ref?: string | null
  client_secret_ref?: string | null
  webhook_secret_ref?: string | null
}

export interface ZoomAccount {
  account_id: string
  provider_kind: string
  display_name: string
  external_account_id: string
  auth_shape: string
  lifecycle_state: string
  runtime_kind: string
  account_email?: string | null
  config: Record<string, unknown>
  created_at: string
  updated_at: string
}

export interface ZoomAccountSetupResponse {
  account: ZoomAccount
}

export interface ZoomAccountListResponse {
  items: ZoomAccount[]
}

export interface ZoomRuntimeStatus {
  account_id: string
  provider_kind: string
  runtime_kind: string
  status: string
  healthy: boolean
  auth_shape: string
  live_runtime_available: boolean
  recording_ingest_available: boolean
  transcript_ingest_available: boolean
  runtime_blockers: string[]
  last_error?: string | null
  checked_at: string
  metadata: Record<string, unknown>
}

export interface ZoomRuntimeStartRequest {
  account_id: string
  force?: boolean
}

export interface ZoomRuntimeStopRequest {
  account_id: string
  reason?: string | null
}

export interface ZoomRuntimeRemoveRequest {
  account_id: string
  reason?: string | null
}

export interface ZoomRuntimeRemoveResponse {
  account_id: string
  provider_kind: string
  removed: boolean
  removed_at: string
}

export interface ZoomCapabilityStatus {
  capability: string
  category: string
  status: string
  action_class: string
  confirmation_required: boolean
  reason: string
}

export interface ZoomCapabilitiesResponse {
  version: string
  runtime_mode: string
  capabilities: ZoomCapabilityStatus[]
  planned_features: string[]
  unsupported_features: string[]
}

export interface ZoomOAuthStartRequest {
  account_id: string
  display_name: string
  external_account_id: string
  account_email?: string | null
  client_id: string
  client_secret?: string | null
  client_secret_ref?: string | null
  webhook_secret_ref?: string | null
  redirect_uri: string
  app_return_url?: string | null
  scopes?: string[]
  authorization_endpoint?: string | null
  token_endpoint?: string | null
  metadata?: Record<string, unknown>
}

export interface ZoomOAuthStartResponse {
  setup_id: string
  authorization_url: string
  state: string
  redirect_uri: string
}

export interface ZoomOAuthCompleteRequest {
  setup_id: string
  state: string
  authorization_code: string
  external_account_id?: string | null
}

export interface ZoomServerToServerAuthorizeRequest {
  account_id: string
  client_id: string
  client_secret?: string | null
  client_secret_ref?: string | null
  zoom_account_id?: string | null
  token_endpoint?: string | null
  metadata?: Record<string, unknown>
}

export interface ZoomAuthorizationResult {
  account_id: string
  provider_kind: string
  auth_shape: string
  lifecycle_state: string
  runtime_kind: string
  token_secret_ref: string
  client_secret_ref?: string | null
  secret_kind: string
  store_kind: string
  authorized_at: string
}

export interface ZoomTokenRefreshRequest {
  account_id: string
  force?: boolean
  refresh_expiring_within_seconds?: number | null
}

export interface ZoomTokenRefreshResult {
  account_id: string
  provider_kind: string
  auth_shape: string
  token_secret_ref: string
  refreshed: boolean
  refresh_strategy: string
  status: string
  expires_at: string
  checked_at: string
  secret_kind: string
  store_kind: string
}

export interface ZoomTokenMaintenanceRequest {
  account_id?: string | null
  force?: boolean
  refresh_expiring_within_seconds?: number | null
}

export interface ZoomTokenMaintenanceItem {
  account_id: string
  provider_kind: string
  auth_shape: string
  status: string
  refreshed: boolean
  expires_at?: string | null
  error?: string | null
}

export interface ZoomTokenMaintenanceResult {
  checked_count: number
  refreshed_count: number
  skipped_count: number
  failed_count: number
  refresh_expiring_within_seconds: number
  checked_at: string
  items: ZoomTokenMaintenanceItem[]
}

export interface ZoomRecordingSyncRequest {
  account_id: string
  user_id?: string | null
  from: string
  to: string
  page_size?: number | null
  max_meetings?: number | null
  api_base_url?: string | null
}

export interface ZoomRecordingSyncFailure {
  meeting_id: string
  step: string
  error: string
}

export interface ZoomRecordingSyncResult {
  account_id: string
  user_id: string
  from: string
  to: string
  meetings_seen: number
  meetings_recorded: number
  recordings_recorded: number
  media_downloads_recorded: number
  transcripts_recorded: number
  failures: ZoomRecordingSyncFailure[]
}

export interface ZoomRecordingImportAuditItem {
  attachment_id: string
  account_id: string
  meeting_id?: string | null
  meeting_uuid?: string | null
  recording_id?: string | null
  filename?: string | null
  content_type: string
  size_bytes: number
  sha256: string
  source?: string | null
  scan_status: string
  scan_summary?: string | null
  storage_kind: string
  storage_path: string
  retention_mode: string
  retention_days: number
  expires_at?: string | null
  created_at: string
}

export interface ZoomRecordingImportAuditResponse {
  account_id: string
  items: ZoomRecordingImportAuditItem[]
}

export interface ZoomRecordingImportRemoveRequest {
  reason?: string | null
}

export interface ZoomRecordingImportRemoveResponse {
  account_id: string
  attachment_id: string
  blob_id: string
  recording_id?: string | null
  removed: boolean
  blob_metadata_removed: boolean
  blob_file_removed: boolean
  removed_at: string
}

export interface ZoomRetentionCleanupRequest {
  remove_recordings?: boolean
  remove_transcripts?: boolean
  limit?: number
}

export interface ZoomRetentionCleanupItem {
  evidence_kind: string
  entity_id: string
  call_id?: string | null
  meeting_id?: string | null
  recording_id?: string | null
  transcript_id?: string | null
  expires_at?: string | null
  removed_at: string
}

export interface ZoomRetentionCleanupResponse {
  account_id: string
  checked_at: string
  recordings_removed: number
  transcripts_removed: number
  items: ZoomRetentionCleanupItem[]
}

export interface ZoomAuditEventItem {
  position: number
  event_id: string
  event_type: string
  occurred_at: string
  subject_kind?: string | null
  subject_entity_id?: string | null
  correlation_id?: string | null
  source: Record<string, unknown>
  subject: Record<string, unknown>
  payload: Record<string, unknown>
  provenance: Record<string, unknown>
}

export interface ZoomAuditEventResponse {
  account_id: string
  items: ZoomAuditEventItem[]
}

export interface ZoomWebhookSubscriptionStatusRequest {
  account_id: string
  api_base_url?: string | null
}

export interface ZoomWebhookSubscriptionReconcileRequest {
  account_id: string
  endpoint_url: string
  subscription_name?: string | null
  event_types?: string[]
  api_base_url?: string | null
}

export interface ZoomWebhookSubscriptionRemoveRequest {
  account_id: string
  subscription_id?: string | null
  api_base_url?: string | null
}

export interface ZoomWebhookSubscription {
  subscription_id: string
  subscription_name: string
  endpoint_url: string
  event_types: string[]
}

export interface ZoomWebhookSubscriptionStatusResult {
  account_id: string
  provider_kind: string
  auth_shape: string
  checked_at: string
  managed_subscription_id?: string | null
  subscriptions: ZoomWebhookSubscription[]
}

export interface ZoomWebhookSubscriptionReconcileResult {
  account_id: string
  provider_kind: string
  auth_shape: string
  status: string
  checked_at: string
  subscription: ZoomWebhookSubscription
}

export interface ZoomWebhookSubscriptionRemoveResult {
  account_id: string
  provider_kind: string
  auth_shape: string
  removed: boolean
  checked_at: string
  subscription_id?: string | null
}

export interface ZoomParticipantSnapshot {
  participant_id?: string | null
  display_name?: string | null
  email?: string | null
  joined_at?: string | null
  left_at?: string | null
  metadata?: Record<string, unknown>
}

export interface ZoomRecordingRef {
  recording_id: string
  recording_type?: string | null
  download_ref?: string | null
  file_extension?: string | null
  file_size_bytes?: number | null
  recorded_at?: string | null
  metadata?: Record<string, unknown>
}

export interface ZoomMeetingObservationRequest {
  observation_id?: string | null
  account_id: string
  meeting_id: string
  meeting_uuid?: string | null
  topic?: string | null
  host_email?: string | null
  join_url?: string | null
  started_at?: string | null
  ended_at?: string | null
  duration_seconds?: number | null
  participants?: ZoomParticipantSnapshot[]
  recording_refs?: ZoomRecordingRef[]
  transcript_ref?: string | null
  metadata?: Record<string, unknown>
  causation_id?: string | null
  correlation_id?: string | null
}

export interface ZoomMeetingIngestResult {
  call_id: string
  account_id: string
  meeting_id: string
  event_id: string
  status: string
}

export interface ZoomRecordingObservationRequest {
  observation_id?: string | null
  account_id: string
  meeting_id: string
  recording: ZoomRecordingRef
  metadata?: Record<string, unknown>
  causation_id?: string | null
  correlation_id?: string | null
}

export interface ZoomRecordingIngestResult {
  account_id: string
  meeting_id: string
  recording_id: string
  event_id: string
  status: string
}

export interface ZoomTranscriptObservationRequest {
  observation_id?: string | null
  transcript_id: string
  account_id: string
  meeting_id: string
  meeting_uuid?: string | null
  source_recording_ref?: string | null
  language_code?: string | null
  transcript_text: string
  segments?: unknown[]
  metadata?: Record<string, unknown>
  causation_id?: string | null
  correlation_id?: string | null
}

export interface ZoomTranscriptIngestResult {
  transcript_id: string
  call_id: string
  account_id: string
  meeting_id: string
  event_id: string
  status: string
}

export interface ZoomTranscriptFileImportRequest {
  observation_id?: string | null
  transcript_id: string
  account_id: string
  meeting_id: string
  meeting_uuid?: string | null
  source_recording_ref?: string | null
  language_code?: string | null
  file_name?: string | null
  content_type?: string | null
  file_text: string
  metadata?: Record<string, unknown>
  causation_id?: string | null
  correlation_id?: string | null
}

export interface ZoomTranscriptFileImportResult extends ZoomTranscriptIngestResult {
  import_format: string
  parsed_segment_count: number
}

export interface ZoomProviderCall {
  call_id: string
  account_id: string
  provider_call_id: string
  provider_chat_id: string
  direction: string
  call_state: string
  started_at?: string | null
  ended_at?: string | null
  transcription_policy_id?: string | null
  metadata: Record<string, unknown>
  created_at: string
  updated_at: string
}

export interface ZoomProviderCallListResponse {
  items: ZoomProviderCall[]
}

export interface ZoomCallTranscript {
  transcript_id: string
  call_id: string
  account_id: string
  provider_chat_id: string
  transcript_status: string
  stt_provider: string
  source_audio_ref?: string | null
  language_code?: string | null
  transcript_text: string
  segments: unknown
  provenance: unknown
  created_at: string
  updated_at: string
}

export interface ZoomCallTranscriptResponse {
  transcript: ZoomCallTranscript | null
}

File: backend/src/application/provider_runtime_services/zoom.rs
use super::*;

impl ZoomProviderRuntimeApplicationService {
    pub(crate) fn new(store: ZoomProviderRuntimeStore) -> Self {
        Self { store }
    }

    pub(crate) async fn setup_fixture_account(
        &self,
        request: &ZoomAccountSetupRequest,
    ) -> Result<ZoomAccountSetupResponse, ZoomError> {
        self.store.setup_fixture_account(request).await
    }

    pub(crate) async fn setup_live_blocked_account(
        &self,
        request: &ZoomLiveAccountSetupRequest,
    ) -> Result<ZoomAccountSetupResponse, ZoomError> {
        self.store.setup_live_blocked_account(request).await
    }

    pub(crate) async fn start_oauth(
        &self,
        request: &ZoomOAuthStartRequest,
    ) -> Result<ZoomOAuthPendingGrant, ZoomError> {
        self.store.start_oauth(request).await
    }

    pub(crate) async fn complete_oauth(
        &self,
        secret_store: &SecretReferenceStore,
        vault: &HostVault,
        pending: ZoomOAuthPendingGrant,
        authorization_code: &str,
        external_account_id: Option<&str>,
    ) -> Result<ZoomAuthorizationResult, ZoomError> {
        self.store
            .complete_oauth(
                secret_store,
                vault,
                pending,
                authorization_code,
                external_account_id,
            )
            .await
    }

    pub(crate) async fn authorize_server_to_server(
        &self,
        secret_store: &SecretReferenceStore,
        vault: &HostVault,
        request: &ZoomServerToServerAuthorizeRequest,
    ) -> Result<ZoomAuthorizationResult, ZoomError> {
        self.store
            .authorize_server_to_server(secret_store, vault, request)
            .await
    }

    pub(crate) async fn refresh_token(
        &self,
        secret_store: &SecretReferenceStore,
        vault: &HostVault,
        request: &ZoomTokenRefreshRequest,
    ) -> Result<ZoomTokenRefreshResult, ZoomError> {
        self.store.refresh_token(secret_store, vault, request).await
    }

    pub(crate) async fn maintain_tokens(
        &self,
        secret_store: &SecretReferenceStore,
        vault: &HostVault,
        request: &ZoomTokenMaintenanceRequest,
    ) -> Result<ZoomTokenMaintenanceResult, ZoomError> {
        self.store
            .maintain_tokens(secret_store, vault, request)
            .await
    }

    pub(crate) async fn list_accounts(
        &self,
        include_removed: bool,
    ) -> Result<ZoomAccountListResponse, ZoomError> {
        self.store.list_accounts(include_removed).await
    }

    pub(crate) async fn runtime_status(
        &self,
        account_id: &str,
    ) -> Result<ZoomRuntimeStatus, ZoomError> {
        self.store.runtime_status(account_id).await
    }

    pub(crate) async fn start_runtime(
        &self,
        request: &ZoomRuntimeStartRequest,
    ) -> Result<ZoomRuntimeStatus, ZoomError> {
        self.store.start_runtime(request).await
    }

    pub(crate) async fn stop_runtime(
        &self,
        request: &ZoomRuntimeStopRequest,
    ) -> Result<ZoomRuntimeStatus, ZoomError> {
        self.store.stop_runtime(request).await
    }

    pub(crate) async fn remove_runtime(
        &self,
        request: &ZoomRuntimeRemoveRequest,
    ) -> Result<ZoomRuntimeRemoveResponse, ZoomError> {
        self.store.remove_runtime(request).await
    }

    pub(crate) async fn observe_meeting(
        &self,
        request: &ZoomMeetingObservationRequest,
    ) -> Result<ZoomMeetingIngestResult, ZoomError> {
        self.store.observe_meeting(request).await
    }

    pub(crate) async fn observe_recording(
        &self,
        request: &ZoomRecordingObservationRequest,
    ) -> Result<ZoomRecordingIngestResult, ZoomError> {
        self.store.observe_recording(request).await
    }

    pub(crate) async fn import_recording_media_download(
        &self,
        request: &ZoomRecordingMediaDownloadRequest,
        bearer_token: Option<&str>,
    ) -> Result<ZoomRecordingMediaImportResult, ZoomError> {
        self.store
            .import_recording_media_download(request, bearer_token)
            .await
    }

    pub(crate) async fn list_recording_imports(
        &self,
        account_id: &str,
        limit: i64,
    ) -> Result<ZoomRecordingImportAuditResponse, ZoomError> {
        self.store.list_recording_imports(account_id, limit).await
    }

    pub(crate) async fn remove_recording_import(
        &self,
        account_id: &str,
        attachment_id: &str,
        request: &ZoomRecordingImportRemoveRequest,
    ) -> Result<ZoomRecordingImportRemoveResponse, ZoomError> {
        self.store
            .remove_recording_import(account_id, attachment_id, request)
            .await
    }

    pub(crate) async fn list_audit_events(
        &self,
        account_id: &str,
        limit: i64,
    ) -> Result<ZoomAuditEventResponse, ZoomError> {
        self.store.list_audit_events(account_id, limit).await
    }

    pub(crate) async fn cleanup_retention(
        &self,
        account_id: &str,
        request: &ZoomRetentionCleanupRequest,
    ) -> Result<ZoomRetentionCleanupResponse, ZoomError> {
        self.store.cleanup_retention(account_id, request).await
    }

    pub(crate) async fn observe_transcript(
        &self,
        request: &ZoomTranscriptObservationRequest,
    ) -> Result<ZoomTranscriptIngestResult, ZoomError> {
        self.store.observe_transcript(request).await
    }

    pub(crate) async fn import_transcript_file(
        &self,
        request: &ZoomTranscriptFileImportRequest,
    ) -> Result<ZoomTranscriptFileImportResult, ZoomError> {
        self.store.import_transcript_file(request).await
    }

    pub(crate) async fn import_transcript_file_download(
        &self,
        request: &ZoomTranscriptFileImportRequest,
        download_url: &str,
        download_token: Option<&str>,
    ) -> Result<ZoomTranscriptFileImportResult, ZoomError> {
        self.store
            .import_transcript_file_download(request, download_url, download_token)
            .await
    }

    pub(crate) async fn sync_recordings(
        &self,
        secret_store: &SecretReferenceStore,
        vault: &HostVault,
        request: &ZoomRecordingSyncRequest,
        allow_remote_recording_downloads: bool,
        allow_remote_transcript_downloads: bool,
    ) -> Result<ZoomRecordingSyncResult, ZoomError> {
        self.store
            .sync_recordings(
                secret_store,
                vault,
                request,
                allow_remote_recording_downloads,
                allow_remote_transcript_downloads,
            )
            .await
    }

    pub(crate) async fn webhook_subscription_status(
        &self,
        secret_store: &SecretReferenceStore,
        vault: &HostVault,
        request: &ZoomWebhookSubscriptionStatusRequest,
    ) -> Result<ZoomWebhookSubscriptionStatusResult, ZoomError> {
        self.store
            .webhook_subscription_status(secret_store, vault, request)
            .await
    }

    pub(crate) async fn reconcile_webhook_subscription(
        &self,
        secret_store: &SecretReferenceStore,
        vault: &HostVault,
        request: &ZoomWebhookSubscriptionReconcileRequest,
    ) -> Result<ZoomWebhookSubscriptionReconcileResult, ZoomError> {
        self.store
            .reconcile_webhook_subscription(secret_store, vault, request)
            .await
    }

    pub(crate) async fn remove_webhook_subscription(
        &self,
        secret_store: &SecretReferenceStore,
        vault: &HostVault,
        request: &ZoomWebhookSubscriptionRemoveRequest,
    ) -> Result<ZoomWebhookSubscriptionRemoveResult, ZoomError> {
        self.store
            .remove_webhook_subscription(secret_store, vault, request)
            .await
    }
}
